internal/cephfs/controllerserver.go

```diff
@@ -36,25 +41,111 @@ type ControllerServer struct {
 	// A map storing all volumes with ongoing operations so that additional operations
 	// for that same volume (as defined by VolumeID/volume name) return an Aborted error
 	VolumeLocks *util.VolumeLocks
+
+	// A map storing all snapshots with ongoing operations so that additional operations
+	// for that same snapshot (as defined by SnapshotID/snapshot name) return an Aborted error
+	SnapshotLocks *util.VolumeLocks
 }
 
 // createBackingVolume creates the backing subvolume and on any error cleans up any created entities
-func (cs *ControllerServer) createBackingVolume(ctx context.Context, volOptions *volumeOptions, vID *volumeIdentifier, secret map[string]string) error {
-	cr, err := util.NewAdminCredentials(secret)
-	if err != nil {
-		return status.Error(codes.InvalidArgument, err.Error())
+func (cs *ControllerServer) createBackingVolume(
+	ctx context.Context,
+	volOptions *volumeOptions,
+	vID *volumeIdentifier,
+	parentVolOpt *volumeOptions,
+	pvID *volumeIdentifier,
+	sID *snapshotIdentifier,
+	cr *util.Credentials) error {
+	var err error
+	if sID != nil {
+		snapID := volumeID(sID.FsSnapshotName)
+		err = cloneSnapshot(ctx, parentVolOpt, cr, volumeID(pvID.FsSubvolName), snapID, volumeID(vID.FsSubvolName), volOptions)
+		if err != nil {
+			return err
+		}
+		defer func() {
+			if err != nil {
+				if dErr := purgeVolume(ctx, volumeID(vID.FsSubvolName), cr, volOptions); dErr != nil {
+					klog.Errorf(util.Log(ctx, "failed to delete volume %s: %v"), vID.FsSubvolName, dErr)
+				}
+			}
+		}()
+		// This is a work around to fix sizing issue for cloned images
+		err = resizeVolume(ctx, volOptions, cr, volumeID(vID.FsSubvolName), volOptions.Size)
+		if err != nil {
+			klog.Errorf(util.Log(ctx, "failed to expand volume %s: %v"), vID.FsSubvolName, err)
+			return status.Error(codes.Internal, err.Error())
+		}
+		var clone = CloneStatus{}
+		clone, err = getcloneInfo(ctx, volOptions, cr, volumeID(vID.FsSubvolName))
+		if err != nil {
+			return status.Error(codes.Internal, err.Error())
+		}
+		if clone.Status.State != cephFSCloneCompleted {
+			return ErrCloneInProgress{err: fmt.Errorf("clone is in progress for %v", vID.FsSubvolName)}
```

Comment:
As err is set, the defer on errors will purge the volume, although it should not?